internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,192 @@
+package graph
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/oklog/ulid/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	parser "github.com/openfga/language/pkg/go/transformer"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/storage/storagewrappers"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+func TestSingleflightResolver(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	var tuples = []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:1", "viewer", "user:jon"),
+
+		tuple.NewTupleKey("folder:2", "parent1", "folder:1"),
+		tuple.NewTupleKey("folder:2", "parent2", "folder:1"),
+
+		tuple.NewTupleKey("folder:3", "parent1", "folder:2"),
+		tuple.NewTupleKey("folder:3", "parent2", "folder:2"),
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	typedefs := parser.MustTransformDSLToProto(`model
```

Comment:
This line and the one below can be collapsed into simply:

```go
model := testutils.MustTransformDSLToProtoWithID(`model
schema 1.1

type user

type folder
relations
	define parent1: [folder]
	define parent2: [folder]
	define viewer: [user] or viewer from parent1 or viewer from parent2 or viewer`)

ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(model))
```

`testutils.MustTransformDSLToProtoWithID(modelString)` is a helper to shortcut that.